### ISSUE TITLE
Native Application still not found, despite copying the pywalfox.json file into the directory from MDN Web Docs, and reinstalling

### ISSUE THREAD

**Issue by VNGamerGit** [Frewacom/pywalfox#77]
This is what the .json file look like:
`{`
  `"name": "pywalfox",`
  `"description": "Automatically theme your browser using the colors generated by Pywal",`
  `"path": "/home/bphuc/.local/lib/python3.9/site-packages/pywalfox/bin/main.sh",`
  `"type": "stdio",`
  `"allowed_extensions": [ "pywalfox@frewacom.org" ]`
`}`
I tried uninstall then reinstall Firefox via Mozilla's website (#14), but it doesn't work. Same for copying the .json file into the directory from MDN Web Docs. I've also tried updating the `path` in the copied manifest, but it still doesn't work
Edit: apparently I copied the wrong file, sorry for opening this without knowing you have to copy the manifest file, not the pywalfox

**Comment by Frewacom** [comment]
Hey, no worries. You can try the steps described here: https://github.com/Frewacom/pywalfox/issues/72#issuecomment-889239885

It will most likely fix your issue 😃

**Comment by VNGamerGit** [comment]
> Hey, no worries. You can try the steps described here: [#72 (comment)](https://github.com/Frewacom/pywalfox/issues/72#issuecomment-889239885)
> 
> It will most likely fix your issue smiley

Still doesn't work, I still got the "Disconnected: version not set" and "Fetch failed: You are not connected to the Pywalfox daemon" error. I have a question though, do you have to set the `path` to the directory Linux Terminal gave you after you ran `pywalfox install` into the `manifest.json` and `pywalfox.json` files in their respective directory, then copy all of them into `/usr/lib/mozilla/native-messaging-hosts`, or do you use the `<path-to-python-site-packages>/pywalfox/bin/main.sh` instead? The `path` the Terminal gave me is `Set executable path in manifest to: /home/<username>/.local/bin/pywalfox`

**Comment by Frewacom** [comment]
> Still doesn't work, I still got the "Disconnected: version not set" and "Fetch failed: You are not connected to the Pywalfox daemon" error. I have a question though, do you have to set the `path` to the directory Linux Terminal gave you after you ran `pywalfox install` into the `manifest.json` and `pywalfox.json` files in their respective directory, then copy all of them into `/usr/lib/mozilla/native-messaging-hosts`, or do you use the `<path-to-python-site-packages>/pywalfox/bin/main.sh` instead? The `path` the Terminal gave me is `Set executable path in manifest to: /home/<username>/.local/bin/pywalfox`

The path is different based on what version of pywalfox you are using. The version I linked in my comment will link to the executable (`/home/<username>/.local/bin/pywalfox`) directly, rather than going through the launch script (`main.sh`). This was the most common cause of the setup issue that many had, since the `PATH` variable would be different for the Firefox executable and your user. 

In most cases, you should not have to do any manual installation of the manifest, nor make changes to it. Looking at your output, it does seem like everything is setup properly, so the issue is probably something else. 

Have you tried running `pywalfox start` in your terminal? Perhaps there is some issue when running the executable? If you have not already, I would also suggest looking in the Browser console for potential errors from pywalfox (`Tools > Web developer > Browser console`).

**Comment by VNGamerGit** [comment]
> > Still doesn't work, I still got the "Disconnected: version not set" and "Fetch failed: You are not connected to the Pywalfox daemon" error. I have a question though, do you have to set the `path` to the directory Linux Terminal gave you after you ran `pywalfox install` into the `manifest.json` and `pywalfox.json` files in their respective directory, then copy all of them into `/usr/lib/mozilla/native-messaging-hosts`, or do you use the `<path-to-python-site-packages>/pywalfox/bin/main.sh` instead? The `path` the Terminal gave me is `Set executable path in manifest to: /home/<username>/.local/bin/pywalfox`
> 
> The path is different based on what version of pywalfox you are using. The version I linked in my comment will link to the executable (`/home/<username>/.local/bin/pywalfox`) directly, rather than going through the launch script (`main.sh`). This was the most common cause of the setup issue that many had, since the `PATH` variable would be different for the Firefox executable and your user.
> 
> In most cases, you should not have to do any manual installation of the manifest, nor make changes to it. Looking at your output, it does seem like everything is setup properly, so the issue is probably something else.
> 
> Have you tried running `pywalfox start` in your terminal? Perhaps there is some issue when running the executable? If you have not already, I would also suggest looking in the Browser console for potential errors from pywalfox (`Tools > Web developer > Browser console`).

I tried running `pywalfox start` in my terminal, but it's not responding, and I got this in the browser console:
`Failed to update option state, target is undefined/null settings.bundle.js:16:9634
    Mt moz-extension://8b93093c-1d53-40e9-86f4-1538d2c6f20b/dist/settings.bundle.js:16
    zt moz-extension://8b93093c-1d53-40e9-86f4-1538d2c6f20b/dist/settings.bundle.js:16
    Jt moz-extension://8b93093c-1d53-40e9-86f4-1538d2c6f20b/dist/settings.bundle.js:16
    Jt moz-extension://8b93093c-1d53-40e9-86f4-1538d2c6f20b/dist/settings.bundle.js:16
    Jt moz-extension://8b93093c-1d53-40e9-86f4-1538d2c6f20b/dist/settings.bundle.js:16
`

**Comment by VNGamerGit** [comment]
> > Still doesn't work, I still got the "Disconnected: version not set" and "Fetch failed: You are not connected to the Pywalfox daemon" error. I have a question though, do you have to set the `path` to the directory Linux Terminal gave you after you ran `pywalfox install` into the `manifest.json` and `pywalfox.json` files in their respective directory, then copy all of them into `/usr/lib/mozilla/native-messaging-hosts`, or do you use the `<path-to-python-site-packages>/pywalfox/bin/main.sh` instead? The `path` the Terminal gave me is `Set executable path in manifest to: /home/<username>/.local/bin/pywalfox`
> 
> The path is different based on what version of pywalfox you are using. The version I linked in my comment will link to the executable (`/home/<username>/.local/bin/pywalfox`) directly, rather than going through the launch script (`main.sh`). This was the most common cause of the setup issue that many had, since the `PATH` variable would be different for the Firefox executable and your user.
> 
> In most cases, you should not have to do any manual installation of the manifest, nor make changes to it. Looking at your output, it does seem like everything is setup properly, so the issue is probably something else.
> 
> Have you tried running `pywalfox start` in your terminal? Perhaps there is some issue when running the executable? If you have not already, I would also suggest looking in the Browser console for potential errors from pywalfox (`Tools > Web developer > Browser console`).

Also, I tried reinstalling pywalfox again, and now it said I need to go through the main script (I used the version you linked)

**Comment by Frewacom** [comment]
`pywalfox start` will run in the background and if you do not receive any errors, it should work as expected. 

Could you check the debugging console of the addon (`about:debugging > This Firefox`, find Pywalfox and click `Inspect > Console` )? 

With the addon console open, you can also try reloading the addon by executing the following: 

```javascript
browser.runtime.reload();
```

**Comment by VNGamerGit** [comment]
> `pywalfox start` will run in the background and if you do not receive any errors, it should work as expected.
> 
> Could you check the debugging console of the addon (`about:debugging > This Firefox`, find Pywalfox and click `Inspect > Console` )?
> 
> With the addon console open, you can also try reloading the addon by executing the following:
> 
> ```js
> browser.runtime.reload();
> ```

wait, where is `about:debugging > This Firefox`?

**Comment by VNGamerGit** [comment]
> `pywalfox start` will run in the background and if you do not receive any errors, it should work as expected.
> 
> Could you check the debugging console of the addon (`about:debugging > This Firefox`, find Pywalfox and click `Inspect > Console` )?
> 
> With the addon console open, you can also try reloading the addon by executing the following:
> 
> ```js
> browser.runtime.reload();
> ```

meh, still got the "Fetch failed: You are not connected to the Pywalfox daemon`" error

**Comment by Frewacom** [comment]
> ![Screenshot 2021-10-25 at 17 26 31](https://user-images.githubusercontent.com/43148755/138737075-08913dba-094c-4406-8568-b616c61b217c.png) im getting the same error but for thunderbird -- everything works perfectly for firefox but i'm getting "Fetch failed: You are not connected to the Pywalfox daemon" on thunderbird

The error that you get indicate that the pywalfox native messenger manifest is not installed. This is what the `pywalfox install` command does. Have you ran that command? If you did, did it execute successfully?

**Comment by Frewacom** [comment]
Sorry for late response, I have been busy lately. 

Have you been able to resolve this @VNGamerGit?

**Comment by VNGamerGit** [comment]
> Sorry for late response, I have been busy lately. 
> 
> Have you been able to resolve this @VNGamerGit? 

Yeah, though it was over a month ago, and I forgot how to solve it already

**Comment by maishams** [comment]
> > ![Screenshot 2021-10-25 at 17 26 31](https://user-images.githubusercontent.com/43148755/138737075-08913dba-094c-4406-8568-b616c61b217c.png) im getting the same error but for thunderbird -- everything works perfectly for firefox but i'm getting "Fetch failed: You are not connected to the Pywalfox daemon" on thunderbird
> 
> The error that you get indicate that the pywalfox native messenger manifest is not installed. This is what the `pywalfox install` command does. Have you ran that command? If you did, did it execute successfully?

it was installed in the right place i dont remember completely what i did but i think i had to manually had the extension v2.0.7 to thunderbird (2.0.6 is the version on the thunderbird store) and then it worked

**Comment by Frewacom** [comment]
Looking at the changes between v2.0.6 and v2.0.7, it should not have any effect on the error that you received, but who knows. I had completely forgotten that Thunderbird does not have the latest version, though. Thanks for the heads up 😃 

Since both of you seem to have resolved your issues, I suppose we can close this?

**Comment by VNGamerGit** [comment]
> Looking at the changes between v2.0.6 and v2.0.7, it should not have any effect on the error that you received, but who knows. I had completely forgotten that Thunderbird does not have the latest version, though. Thanks for the heads up 😃 
> 
> Since both of you seem to have resolved your issues, I suppose we can close this?

Yeah, I suppose. Or maybe we can keep this open in case anyone got the same problem as mine?

**Comment by dave-cao** [comment]
Hey I just wanted to put in a note. I had the same error of "You are not connected to the Pywalfox daemon" and the above methods didn't work.

For others that run into the same problem, the following solution worked for me.

It turned out that I added the pywalfox firefox extension before installing pywalfox on the system. I just had to delete and re-add the pywalfox extension on firefox and it worked.

So in short, `pip install pywalfox` and `pywalfox install` **before** adding in the firefox pywalfox extension. 

Hope this helps anyone in the future.

**Comment by MNolan147** [comment]
I am currently experiencing this bug and none of the above fixes are working for me.

**Comment by phurgelisherpa** [comment]
Reboot your pc, and do `sudo pywalfox install` then try to fetch pywalfox extension
It works for me.

**Comment by MNolan147** [comment]
This didn't work for me. The only difference is that the pywalfox extension no longer appears in the extension bar (it is installed but I can't get to the page that allows me to fetch colours).

**Comment by phurgelisherpa** [comment]
I had that problem too ( extension bar  no longer appear ) and private window not working  ,
Do update your firefox if it didn't work then try reinstall firefox, It fixed that both problem for me

**Comment by Glesas** [comment]
I'm on flatpak version of Zen-browser, and altough my default firefox is working as it should, no matter what folder i'd give access to, or what setting i'd toggle, nothing is working, and zen-browser can't reach the pywalfox daemon. 

I've even tried to toggle absolutley every switch on, which now created separate profile of browser somewhere (not that important), i've installed pywalfox extension on it aswell, and it was simply "None" error. While previously it would've been "No such native application pywalfox".

Here's the view of flatseal, and what i've toggled, to keep my original profile folder with all tabs and bits working as it should, while not breaking zen-browser completley. 
![image](https://github.com/user-attachments/assets/fadc733d-0d95-4cb7-a138-b0f1668dbf6c)
.   I could use the AppImage version of zen-browser, and it does indeed work, but I kinda lowkey don't wanna to manually update the zen-browser. Any other browser - I'm fine with, but zen-browser pretty new and need's polishing, so i wanna to have as many updates as it's going to do. If anyone would find a solution what i should do inside of Flatseal, to make Zen-Browser Flatpak version to work with pywalfox, please let me know.

**Comment by technology-maen** [comment]
I fixed the you are not connected but it it says it can't find colors.json 
To fix the issue 

> go to flatseal and add the path to installed pywal 

> make sure it is present in native messages host json

**Comment by birchtree25** [comment]
> Hey I just wanted to put in a note. I had the same error of "You are not connected to the Pywalfox daemon" and the above methods didn't work.
> 
> For others that run into the same problem, the following solution worked for me.
> 
> It turned out that I added the pywalfox firefox extension before installing pywalfox on the system. I just had to delete and re-add the pywalfox extension on firefox and it worked.
> 
> So in short, `pip install pywalfox` and `pywalfox install` **before** adding in the firefox pywalfox extension.
> 
> Hope this helps anyone in the future.

I am having the same issue and doing it this has worked, just with pipx. Thanks for the help

**Comment by johanengstrand** [comment]
Installation with `pipx` is now suggested in the README, along with other fixes (also related to Flatpaks) in the newest version.
